storage-node/src/storage_manager.rs

```diff
@@ -16,41 +12,52 @@ use storage_common::RequestParams;
 /// which should be responsible for handling multiple requests at the
 /// same time.
 
-pub struct StorageManager<C: DataStoreCache, DS: DataStore> {
-    /// Cache_key = S3_PATH; Cache_value = (CACHE_BASE_PATH + S3_PATH, size)
-    data_store_cache: RwLock<C>,
-    /// We don't use lock here because `data_store` itself should handle the concurrency.
-    data_store: DS,
+pub struct StorageManager<C: DataStoreCache> {
+    /// We don't use lock here because `data_store_cache` itself should handle the concurrency.
+    data_store_cache: C,
 }
 
-impl<C: DataStoreCache, DS: DataStore> StorageManager<C, DS> {
-    pub fn new(data_store_cache: C, data_store: DS) -> Self {
-        Self {
-            data_store_cache: RwLock::new(data_store_cache),
-            data_store,
-        }
+impl<C: DataStoreCache> StorageManager<C> {
+    pub fn new(data_store_cache: C) -> Self {
+        Self { data_store_cache }
     }
 
-    fn dummy_s3_request(&self) -> (String, Vec<String>) {
-        let bucket = "tests-parquet".to_string();
-        let keys = vec![
-            "userdata1.parquet".to_string(),
-            "userdata2.parquet".to_string(),
-        ];
+    // In dummy function, we assume request = bucket:key0+key1+...
+    // TODO(lanlou): When we change dummy_s3_request, we should also change tests
+    fn dummy_s3_request(&self, request: String) -> (String, Vec<String>) {
+        let mut iter = request.split(':');
+        let bucket = iter.next().unwrap().to_string();
+        let keys = iter
+            .next()
+            .unwrap()
+            .split('+')
+            .map(|s| s.to_string())
+            .collect();
         (bucket, keys)
     }
 
-    pub async fn get_data(&self, _request: RequestParams) -> ParpulseResult<usize> {
+    pub async fn get_data(&mut self, request: RequestParams) -> ParpulseResult<usize> {
         // 1. Try to get data from the cache first.
         // 2. If cache miss, then go to storage reader to fetch the data from
         // the underlying storage.
         // 3. If needed, update the cache with the data fetched from the storage reader.
         // TODO: Support more request types.
-        let (bucket, keys) = self.dummy_s3_request();
+        let request = match request {
+            RequestParams::S3(request) => request,
+            _ => {
+                return Err(ParpulseError::Internal(
+                    "Unsupported request types".to_string(),
+                ))
+            }
+        };
+        let (bucket, keys) = self.dummy_s3_request(request.clone());
 
         // FIXME: Cache key should be <bucket + key>. Might refactor the underlying S3
         // reader as one S3 key for one reader.
-        let data_rx = self.get_data_from_cache(bucket.clone()).await?;
+        let data_rx = self
+            .data_store_cache
+            .get_data_from_cache(request.clone())
+            .await?;
```

Comment:
Here we pass the request string directly into `get_data_from_cache`, which is assumed to be `bucket:key0+key1+...` according to the code comment above. But the method in fact accepts a single S3 key (i.e. `bucket key0`). 
Currently the `S3ReaderStream` accepts multiple keys, so polling data from an `S3ReaderStream` will yield data for `bucket:key0+key1+...` instead of `bucket key0` (though we can achieve the same effect by constructing several `S3ReaderStream`, each having a single S3 key). We should further discuss the behavior of the storage manager, the cache and the S3 stream to make them consistent.